### ISSUE TITLE
Logging of all config items with sources

### ISSUE THREAD

**Issue by adamralph** [config-r/config-r#19]
Using JSV serialization.

```
[INFO]   Foo: {Bar:bar,Baz:baz} (MyApp.exe.csx)
[DEBUG]    Foo: {Bar:barr,Baz:bazz} (Custom.csx)
[DEBUG]    Foo: {Bar:barrr,Baz:bazzz} (Custom2.csx)
[INFO]   MaxRedirects: 3000 (Custom.csx)
[DEBUG]    MaxRedirects: 999 (Custom2.csx)
```

**Comment by teabaggs** [comment]
Hello Adam,
is this meant to be done at load time or on demand? should it default to on or off? Can you explain why some Foos are [INFO] and some are [DEBUG] level? I am imagining all logs of this sort are [DEBUG] level and user can effectively enable/disable by adjusting the log level.

**Comment by adamralph** [comment]
hi @teabaggs 

The initial idea was to do this on demand, but now you've got me thinking. It's probably better just to do this implicitly and, as you say, let the user control the output by tuning the config of their logging framework of choice.

The reason that some are INFO and some are DEBUG is that the INFO level ones are the final values of each key, and the DEBUG ones are the masked ones, loaded by nested configs.

E.g. the above log would be generated by:

```csharp
// MyApp.exe.csx
LoadLoadScriptFile("Custom.csx"),
Add("Foo", new Foo { Bar = "bar", Baz = "baz" });
```
```csharp
// Custom.csx
LoadLoadScriptFile("Custom2.csx"),
Add("Foo", new Foo { Bar = "barr", Baz = "bazz" });
Add("MaxRedirects", 3000 });
```
```csharp
// Custom2.csx
Add("Foo", new Foo { Bar = "barrr", Baz = "bazzz" });
Add("MaxRedirects", 999 });
```

**Comment by cmatskas** [comment]
@adamralph what logging framework are you using? Do you want to use abstractions such as LibLog (my favorite at the moment) or Common.Logging to leave the responsibility to the user to add a concrete logger implementation (nLog etc). My suggestions would be:

- Add an abstract logging framework to ConfigR
- Provide settings to switch it on/off
- Let users implement the logger

**Comment by adamralph** [comment]
@cmatskas currently ConfigR is using Common.Logging since it inherited the reference from scriptcs.

However, scriptcs [deprecated Common.Logging](http://blog.scriptcs.net/2015/08/03/deprecating-common-logging/) in version 0.15 so I think it makes sense to first do an isolated refactoring PR containing purely a switch from Common.Logging to LibLog and nothing more. Do you fancy taking that on? Note that ConfigR will still inherit the Common.Logging reference from scriptcs, since it has only been deprecated and not removed.

**Comment by cmatskas** [comment]
@adamralph lol, I was hoping(fearing) this was coming my way. You know what? I'll take this one :)
I'll do the PR to switch from using Common.Logging to LibLog. Should that be a separate issue to keep better track of things?

**Comment by adamralph** [comment]
Thanks dude! Don't worry about raising a separate issue just yet. Let's see
what the effect of the change is first. We can always raise one later if we
want.

**Comment by cmatskas** [comment]
Morning @adamralph. The replacement of Common.Logging is going well, however there are a couple of places where we need to use a Common.Logging logger for the ScriptCs stuff. Are you happy to keep both for the purpose of ScriptCs only?

**Comment by adamralph** [comment]
Hi @cmatskas the Common.Logging overloads have been deprecated in scriptcs. You should be able to use the overloads which take a `ScriptCs.Contracts.ILogProvider` and provide `ScriptCs.Contracts.DefaultLogProvider`.

**Comment by cmatskas** [comment]
Cool, I'll sort it out!

CM

-----Original Message-----

**Comment by cmatskas** [comment]
Hi @adamralph. I don't see the deprecation in of Common.Logging in ScriptCs. Looking into the ScriptExecutor class I can only see 2 constructors:
    
    public ScriptExecutor(IFileSystem fileSystem, IFilePreProcessor filePreProcessor, IScriptEngine scriptEngine, ILog logger);
    public ScriptExecutor(IFileSystem fileSystem, IFilePreProcessor filePreProcessor, IScriptEngine scriptEngine, ILog logger, IScriptLibraryComposer composer);

Where is the overload you're talking about? thanks for your help

**Comment by adamralph** [comment]
@cmatskas it's because ConfigR is still using scriptcs 0.14. You'll have to 0.15 first. That should be a separate commit.

**Comment by cmatskas** [comment]
Arg! Lot's of your NuGet packages are out of date :( Do you want me to do them all in one go?
And yes, I found out why it wasn't there :)

**Comment by adamralph** [comment]
For the packages used in tests only, sure, feel free to upgrade them too. For those which are runtime dependencies, I'd rather leave them for now. There's no use in forcing people up to newer versions unnecessarily.

**Comment by cmatskas** [comment]
Right. In that case, I'll upgrade ScriptCs as needed for the LibLog upgrade and all packages used in tests. This will be a separate branch and a separate PR. Thanks (what did I get myself into? lol)

**Comment by adamralph** [comment]
That sounds ideal. All very much appreciated!

**Comment by cmatskas** [comment]
Lol, I'm enjoying it :)

-----Original Message-----

**Comment by adamralph** [comment]
That's the most important thing :wink:

**Comment by cmatskas** [comment]
@adamralph The upgrade of NuGet packages, especially the ScriptCs part, can't happen as a separate check in because version 0.15 removes the dependency to Common.Logging and throws a warning as an error. It will have to be done in one PR I'm afraid. Would you be happy with that?

**Comment by adamralph** [comment]
There must be something wrong. scriptcs 0.15 still has a dep on
Common.Logging. Anyhow, you can just reinstall it yourself.

**Comment by cmatskas** [comment]
Ah, the upgrade of the package is fine. However when I compile the code (without any changes in) I get a Warning as Error compilation error because `Support for Common.Logging types was deprecated in version 0.15.0`. Thouhgts?

**Comment by adamralph** [comment]
Ah, there are suppressions you can put into the csproj. You should be able
to find an example in one of the scriptcs csproj files.

**Comment by cmatskas** [comment]
Cool, as long as we are aware of this, I'm good :) Thanks for quick feedback!

**Comment by teabaggs** [comment]
Hi @adamralph . I'm back on task after some delay. I've been thinking about this and I have an idea which I'd like to discuss. These are to enable finding the actual source of a config value when logging. Cascade() can take the appropriate value from a dictionary but it cannot tell which source (file, web script, etc. ) this value came from. So in order to log the source we have to track this. How to do this? One idea is to extend ISimpleConfig to include 
```
string Source { get; }
```
and promote Source to public on BasicConfig and it's sub-classes. This would allow us to get the source string for a given config and log it.

### Log Strategy

Now back to the question of on demand or implicitly logging this information. If we do it implicitly it should, IMHO, be done in 
```
public IConfig Load(ISimpleConfig config)
```
since this is when a key/value might be overwritten or added. If a program loads from many config sources this seems like unnecessary overhead since only the last load matters for the final values. So it could be moved to
```
public IConfig EnsureLoaded(params Assembly[] references)
```
but this would require calling the method if global auto loading is disabled, which makes it explicit sometimes and implicit other times.

So I propose a new LogStrategy enum
```
  public enum LogStrategy
    {
        OnLoad,
        OnDemand,
    }
```
and a check in Load method
```
   public IConfig Load(ISimpleConfig config)
        {
            this.loadInvoked = true;

            Guard.AgainstNullArgument("config", config);

            var index = this.dictionaries.Count;
            this.dictionaries.Add(config);
            try
            {
                config.Load();
            }
            catch
            {
                this.dictionaries.RemoveRange(index, this.dictionaries.Count - index);
                throw;
            }

            if (LoggingStrategy == LogStrategy.OnLoad)
            {
                LogConfigSources();
            }
            return this;
        }
```
where the LogConfigSources method is responsible for writing the logs. It could be public and so could be implicitly or explicitly called depending on the LoggingStrategy.

There are of course other options but after consideration I feel this is the cleanest since Source is implemented (but not public) and is a reasonable api for enabling/disabling logging programatically as well as with the log level. Maybe I'm overthinking it or missing something. I'm interested to know what you think.

-- Cheers

**Comment by cmatskas** [comment]
I like it and it looks good. My +1 on this

**Comment by adamralph** [comment]
ConfigR is being re-written. This idea will be considered during that re-write and the status of this issue will be reconsidered after the re-write is complete.

**Comment by adamralph** [comment]
After #246 this is invalid. ConfigR no longer controls the context of which file is adding a config item, since this is all delegated to Roslyn's `#load`.

If someone wants to log the config items after loading them, they can use their favourite logging library to do this. That's not a concern of ConfigR.
